### PR TITLE
Finalize single worker fscache support

### PR DESCRIPTION
**Pull request by hsiangkao**
Step I. prepare a json file:
{
  "type": "bootstrap",
  "id": "aa229b3a8eeeacf0bfbb17a6415fa6a817583d99c4b590fc6b56a50f49e8957c",
  "domain_id": "aa229b3a8eeeacf0bfbb17a6415fa6a817583d99c4b590fc6b56a50f49e8957c",
  "config": {
    "id": "factory1",
    "backend_type": "registry",
    "backend_config": {
      "scheme": "https",
      "host": "index.docker.io",
      "repo": "hsiangkao/ubuntu"
    },
    "cache_type": "fscache",
    "cache_config": {
      "work_dir": "/var/lib/containerd-nydus-grpc/cachedir"
    },
    "metadata_path": "/root/ubuntu-20.04.image.boot"
  }
}

Step II.
docker poll docker.io/hsiangkao/ubuntu:20.04-rafs-v6-docker
and copy image.boot as /root/ubuntu-20.04.img.boot

Step III.
cargo build --target x86_64-unknown-linux-gnu --features=fusedev --release --target-dir target-fusedev --bin nydusd
mkdir /root/fscache
target-fusedev/x86_64-unknown-linux-gnu/release/nydusd daemon -F /root/fscache -A /tmp/nydus.sock

Step IV.
curl --http1.1 --unix-socket /tmp/nydus.sock -XPUT -d "@[json file]" http://localhost/api/v2/blobs
mount -t erofs -ofsid=aa229b3a8eeeacf0bfbb17a6415fa6a817583d99c4b590fc6b56a50f49e8957c,device=aae220288dc01b26a0d69ef553a32d274b258aff545d798e04f91045e00b3666 nodev mnt

Signed-off-by: Gao Xiang \<hsiangkao@linux.alibaba.com\>

### CODE REVIEW

File: src/bin/nydusd/fs_cache.rs [hunk 1]
```diff
@@ -478,53 +489,57 @@ impl FsCacheHandler {
...
-            Some(FsCacheObject::DataBlob(blob)) => {
-                let config = state.fd_to_config_map.remove(&msg.fd).unwrap();
-                BLOB_FACTORY.gc(Some((config.factory_config(), blob.blob_id())));
-                format!("cfini {}", hdr.id)
-            }
-        };
-        self.reply(&msg);
+
+        if let Some(FsCacheObject::DataBlob(blob)) = state.fd_to_object_map.remove(&msg.fd) {
+            let blob_id = &blob.blob_id().to_string();
```
Comment (jiangliu) [inline src/bin/nydusd/fs_cache.rs:538]: This change should be unnecessary now

File: src/bin/nydusd/blob_cache.rs [hunk 2]
```diff
@@ -179,6 +179,7 @@ impl BlobCacheMgr {
                 .insert(config.get_key().to_string(), config);
             // Try to add the referenced data blob object if it doesn't exist yet.
             for bi in rs.superblock.get_blob_infos() {
+                info!("Found blob {} on domain {}", &bi.blob_id(), domain_id);
```
Comment (jiangliu) [inline src/bin/nydusd/blob_cache.rs:182]: Should we remove this info? Or change it to debug?

Comment (hsiangkao) [reply src/bin/nydusd/blob_cache.rs:182]: changing info debug makes sense to me, will do.

Comment (hsiangkao) [reply src/bin/nydusd/blob_cache.rs]: done.

File: src/bin/nydusd/fs_cache.rs (hunk 1 above)
Comment (hsiangkao) [reply src/bin/nydusd/fs_cache.rs]: done.

File: src/bin/nydusd/fs_cache.rs [hunk 3]
```diff
@@ -428,13 +436,16 @@ impl FsCacheHandler {
             Ok((blob, blob_size)) => {
                 let mut state = self.state.lock().unwrap();
                 if state.fd_to_object_map.contains_key(&msg.fd) {
-                    format!("cinit {},{}", hdr.id, -libc::EALREADY)
+                    unsafe {
+                        libc::close(msg.fd as i32);
+                    }
```
Comment (jiangliu) [inline src/bin/nydusd/fs_cache.rs:441]: When dropping the blob object, the msg.fd should be closed already, seems we should not close fd here.

Comment (hsiangkao) [reply src/bin/nydusd/fs_cache.rs]: fixed, thanks!

**Pull request merged**